src/gateway.c

```diff
@@ -1286,6 +1286,55 @@ static int handle_weight(struct handle *req)
 	return 0;
 }
 
+static void leaderGenexecCb(struct exec *exec, int status)
+{
+	struct gateway *g = exec->data;
+	struct handle *req = g->req;
+	sqlite3_stmt *stmt = g->stmt;
+
+	g->req = NULL;
+	g->stmt = NULL;
+
+	if (status != SQLITE_DONE && status != SQLITE_ROW) {
+		assert(g->leader != NULL);
+		failure(req, status, error_message(g->leader->conn, status));
+		sqlite3_reset(stmt);
+		return;
+	}
+	query_batch(stmt, req, status);
+}
+
+static int handle_genexec(struct handle *req)
+{
+	struct cursor *cursor = &req->cursor;
+	struct gateway *g = req->gateway;
+	struct stmt *stmt;
+	uint64_t req_id;
+	int rv;
+
+	START_V0(genexec, rows);
+	(void)response;
+	CHECK_LEADER(req);
+	LOOKUP_DB(request.db_id);
+	LOOKUP_STMT(request.stmt_id);
+	FAIL_IF_CHECKPOINTING;
+	rv = bind__params(stmt->stmt, cursor, TUPLE__PARAMS32);
+	if (rv != 0) {
+		failure(req, rv, "bind parameters");
+		return 0;
+	}
+	g->req = req;
+	g->stmt = stmt->stmt;
+	req_id = idNext(&g->random_state);
+	rv = leader__exec(g->leader, &g->exec, stmt->stmt, req_id, leaderGenexecCb);
```

Comment:
Unless I'm missing something, the issue I see with this approach is that if there is a write-statement in progress (say an `INSERT` or `UPDATE`) which is being committed to Raft, then I believe that any attempt to issue a read-only statement (say a `SELECT`) will fail with `SQLITE_BUSY`, because `leader__exec` can't be executed concurrently, see https://github.com/canonical/dqlite/blob/57d7590d445baae419e2c3119342fe65c1926f79/src/leader.c#L422.

I believe it'd better to use [`sqlite_stmt_readonly`](https://www.sqlite.org/c3ref/stmt_readonly.html) to determine if `leader__exec` is needed or not (this should also avoid the need of the `initial` flag`).

Please also consider the [suggestion](https://github.com/canonical/dqlite/issues/471#issuecomment-1422416138) I made in the issue, to include a new flag in the `PREPARE` response and let the client drive the conversation.

Reply:
That should work.